File: workers/dotnet/projects/harness/Worker.cs
using System.CommandLine;
using System.CommandLine.Invocation;
using System.CommandLine.Parsing;
using System.Runtime.ExceptionServices;
using Microsoft.Extensions.Logging;
using Temporalio.Client;
using Temporalio.Worker;
using Temporalio.Worker.Tuning;

namespace Temporalio.Omes.Projects.Harness;

public sealed record WorkerContext(
    ILogger Logger,
    string TaskQueue,
    bool ErrOnUnimplemented,
    TemporalWorkerOptions WorkerOptions);

public delegate TemporalWorker WorkerFactory(ITemporalClient client, WorkerContext context);

// This small generic seam exists only to let tests substitute fake worker objects, which is the
// closest .NET analogue to Python patching `_run_workers`.
internal delegate TWorker WorkerFactoryCore<out TWorker>(ITemporalClient client, WorkerContext context);

internal static class WorkerHarness
{
    private static readonly Option<string> ServerAddressOption = new(
        aliases: new[] { "-a", "--server-address" },
        description: "The host:port of the server",
        getDefaultValue: () => "localhost:7233");

    private static readonly Option<string> TaskQueueOption = new(
        aliases: new[] { "-q", "--task-queue" },
        description: "The task queue to use",
        getDefaultValue: () => "omes");

    private static readonly Option<string> NamespaceOption = new(
        aliases: new[] { "-n", "--namespace" },
        description: "The namespace to use",
        getDefaultValue: () => "default");

    private static readonly Option<int> TaskQueueSuffixStartOption = new(
        name: "--task-queue-suffix-index-start",
        description: "Inclusive start for task queue suffix range");

    private static readonly Option<int> TaskQueueSuffixEndOption = new(
        name: "--task-queue-suffix-index-end",
        description: "Inclusive end for task queue suffix range");

    private static readonly Option<int?> MaxConcurrentActivityPollersOption = new(
        name: "--max-concurrent-activity-pollers",
        description: "Max concurrent activity pollers");

    private static readonly Option<int?> MaxConcurrentWorkflowPollersOption = new(
        name: "--max-concurrent-workflow-pollers",
        description: "Max concurrent workflow pollers");

    private static readonly Option<int?> ActivityPollerAutoscaleMaxOption = new(
        name: "--activity-poller-autoscale-max",
        description: "Max for activity poller autoscaling (overrides max-concurrent-activity-pollers)");

    private static readonly Option<int?> WorkflowPollerAutoscaleMaxOption = new(
        name: "--workflow-poller-autoscale-max",
        description: "Max for workflow poller autoscaling (overrides max-concurrent-workflow-pollers)");

    private static readonly Option<int?> MaxConcurrentActivitiesOption = new(
        name: "--max-concurrent-activities",
        description: "Max concurrent activities");

    private static readonly Option<int?> MaxConcurrentWorkflowTasksOption = new(
        name: "--max-concurrent-workflow-tasks",
        description: "Max concurrent workflow tasks");

    private static readonly Option<double?> ActivitiesPerSecondOption = new(
        name: "--activities-per-second",
        description: "Per-worker activity rate limit");

    private static readonly Option<string> LogLevelOption = new Option<string>(
        name: "--log-level",
        description: "Log level",
        getDefaultValue: () => "info");

    private static readonly Option<string> LogEncodingOption = new Option<string>(
        name: "--log-encoding",
        description: "Log encoding",
        getDefaultValue: () => "console").FromAmong("console", "json");

    private static readonly Option<bool> TlsOption = new(
        name: "--tls",
        description: "Enable TLS");

    private static readonly Option<string> TlsCertPathOption = new(
        name: "--tls-cert-path",
        description: "Path to a client certificate for TLS",
        getDefaultValue: () => string.Empty);

    private static readonly Option<string> TlsKeyPathOption = new(
        name: "--tls-key-path",
        description: "Path to a client key for TLS",
        getDefaultValue: () => string.Empty);

    private static readonly Option<string?> PromListenAddressOption = new(
        name: "--prom-listen-address",
        description: "Prometheus listen address");

    private static readonly Option<string> AuthHeaderOption = new(
        name: "--auth-header",
        description: "Authorization header value",
        getDefaultValue: () => string.Empty);

    private static readonly Option<string?> PromHandlerPathOption = new(
        name: "--prom-handler-path",
        description: "Prometheus handler path",
        getDefaultValue: () => "/metrics");

    private static readonly Option<string?> BuildIdOption = new(
        name: "--build-id",
        description: "Build ID");

    private static readonly Option<bool> ErrOnUnimplementedOption = new(
        name: "--err-on-unimplemented",
        description: "Error when receiving unimplemented actions (currently only affects concurrent client actions)",
        getDefaultValue: () => false);

    public static async Task<int> RunAsync(App app, string[] args)
    {
        var command = CreateWorkerCommand();
        command.SetHandler(
            async (InvocationContext context) =>
            {
                try
                {
                    if (context.ParseResult.Errors.Count > 0)
                    {
                        throw new ArgumentException(context.ParseResult.Errors[0].Message);
                    }

                    var options = ParseWorkerOptions(context.ParseResult);
                    await RunCoreAsync(
                        workerFactory: (client, workerContext) => app.Worker(client, workerContext),
                        clientFactory: app.ClientFactory,
                        options: options,
                        runWorkersAsync: RunWorkersAsync);
                    context.ExitCode = 0;
                }
                catch (ArgumentException err)
                {
                    Console.Error.WriteLine(err.Message);
                    context.ExitCode = 1;
                }
            });
        return await command.InvokeAsync(args);
    }

    internal static async Task RunCoreAsync<TWorker>(
        WorkerFactoryCore<TWorker> workerFactory,
        ClientFactory clientFactory,
        WorkerCliOptions options,
        Func<IReadOnlyList<TWorker>, Task> runWorkersAsync)
    {
        if (options.TaskQueueSuffixIndexStart > options.TaskQueueSuffixIndexEnd)
        {
            throw new ArgumentException("Task queue suffix start after end");
        }

        var loggerFactory = HarnessHelpers.ConfigureLogger(options.LogLevel, options.LogEncoding);
        var logger = loggerFactory.CreateLogger("Temporalio.Omes.Projects.Harness");
        var clientConfig = ClientHelpers.BuildClientConfig(
            serverAddress: options.ServerAddress,
            @namespace: options.Namespace,
            authHeader: options.AuthHeader,
            tls: options.Tls,
            tlsCertPath: options.TlsCertPath,
            tlsKeyPath: options.TlsKeyPath,
            promListenAddress: options.PromListenAddress,
            loggerFactory: loggerFactory);

        var workerSettings = BuildWorkerSettings(options);
        var client = await clientFactory(clientConfig);
        var taskQueues = BuildTaskQueues(
            logger,
            options.TaskQueue,
            options.TaskQueueSuffixIndexStart,
            options.TaskQueueSuffixIndexEnd);

        var workers = taskQueues
            .Select(taskQueue =>
                workerFactory(
                    client,
                    new WorkerContext(
                        Logger: logger,
                        TaskQueue: taskQueue,
                        ErrOnUnimplemented: options.ErrOnUnimplemented,
                        WorkerOptions: BuildWorkerOptions(taskQueue, workerSettings))))
            .ToList();

        await runWorkersAsync(workers);
    }

    internal static Task RunWorkersAsync(IReadOnlyList<TemporalWorker> workers) =>
        RunWorkersAsync(
            workers,
            static (worker, cancellationToken) => worker.ExecuteAsync(cancellationToken),
            static worker => worker.Dispose());

    internal static async Task RunWorkersAsync<TWorker>(
        IReadOnlyList<TWorker> workers,
        Func<TWorker, CancellationToken, Task> executeWorkerAsync,
        Action<TWorker> disposeWorker)
    {
        using var cancellationTokenSource = new CancellationTokenSource();
        ConsoleCancelEventHandler? cancelHandler = (_, eventArgs) =>
        {
            eventArgs.Cancel = true;
            cancellationTokenSource.Cancel();
        };

        Console.CancelKeyPress += cancelHandler;
        try
        {
            var workerTasks = workers
                .Select(worker => executeWorkerAsync(worker, cancellationTokenSource.Token))
                .ToList();

            var completedTask = await Task.WhenAny(workerTasks);
            cancellationTokenSource.Cancel();

            Exception? firstFailure = null;
            if (completedTask.IsFaulted)
            {
                firstFailure = completedTask.Exception?.GetBaseException();
            }
            try
            {
                await Task.WhenAll(workerTasks);
            }
            catch
            {
                firstFailure ??= workerTasks
                    .Where(task => task.IsFaulted)
                    .Select(task => task.Exception?.GetBaseException())
                    .FirstOrDefault(exception => exception is not null);
            }

            if (firstFailure is not null)
            {
                ExceptionDispatchInfo.Capture(firstFailure).Throw();
            }
        }
        finally
        {
            Console.CancelKeyPress -= cancelHandler;
            foreach (var worker in workers)
            {
                disposeWorker(worker);
            }
        }
    }

    private static List<string> BuildTaskQueues(
        ILogger logger,
        string taskQueue,
        int suffixStart,
        int suffixEnd)
    {
        if (suffixEnd == 0)
        {
            logger.LogInformation(".NET worker will run on task queue {TaskQueue}", taskQueue);
            return new List<string> { taskQueue };
        }

        var taskQueues = Enumerable.Range(suffixStart, suffixEnd - suffixStart + 1)
            .Select(index => $"{taskQueue}-{index}")
            .ToList();
        logger.LogInformation(".NET worker will run on {Count} task queue(s)", taskQueues.Count);
        return taskQueues;
    }

    private static WorkerSettings BuildWorkerSettings(WorkerCliOptions options) =>
        new(
            MaxConcurrentActivities: options.MaxConcurrentActivities,
            MaxConcurrentWorkflowTasks: options.MaxConcurrentWorkflowTasks,
            ActivitiesPerSecond: options.ActivitiesPerSecond,
            MaxConcurrentActivityPollers: options.MaxConcurrentActivityPollers,
            MaxConcurrentWorkflowPollers: options.MaxConcurrentWorkflowPollers,
            ActivityPollerAutoscaleMax: options.ActivityPollerAutoscaleMax,
            WorkflowPollerAutoscaleMax: options.WorkflowPollerAutoscaleMax);

    private static TemporalWorkerOptions BuildWorkerOptions(string taskQueue, WorkerSettings settings)
    {
        var workerOptions = new TemporalWorkerOptions(taskQueue);
        if (settings.MaxConcurrentWorkflowTasks is { } maxConcurrentWorkflowTasks)
        {
            workerOptions.MaxConcurrentWorkflowTasks = maxConcurrentWorkflowTasks;
        }

        if (settings.MaxConcurrentActivities is { } maxConcurrentActivities)
        {
            workerOptions.MaxConcurrentActivities = maxConcurrentActivities;
        }

        if (settings.ActivitiesPerSecond is { } activitiesPerSecond)
        {
            workerOptions.MaxActivitiesPerSecond = activitiesPerSecond;
        }

        if (settings.ActivityPollerAutoscaleMax is { } activityAutoscaleMax)
        {
            workerOptions.ActivityTaskPollerBehavior = new PollerBehavior.Autoscaling(maximum: activityAutoscaleMax);
        }
        else if (settings.MaxConcurrentActivityPollers is { } maxConcurrentActivityPollers)
        {
            workerOptions.MaxConcurrentActivityTaskPolls = maxConcurrentActivityPollers;
        }

        if (settings.WorkflowPollerAutoscaleMax is { } workflowAutoscaleMax)
        {
            workerOptions.WorkflowTaskPollerBehavior = new PollerBehavior.Autoscaling(maximum: workflowAutoscaleMax);
        }
        else if (settings.MaxConcurrentWorkflowPollers is { } maxConcurrentWorkflowPollers)
        {
            workerOptions.MaxConcurrentWorkflowTaskPolls = maxConcurrentWorkflowPollers;
        }

        return workerOptions;
    }

    internal static WorkerCliOptions ParseWorkerOptions(ParseResult parseResult) =>
        new(
            ServerAddress: parseResult.GetValueForOption(ServerAddressOption)!,
            TaskQueue: parseResult.GetValueForOption(TaskQueueOption)!,
            Namespace: parseResult.GetValueForOption(NamespaceOption)!,
            TaskQueueSuffixIndexStart: parseResult.GetValueForOption(TaskQueueSuffixStartOption),
            TaskQueueSuffixIndexEnd: parseResult.GetValueForOption(TaskQueueSuffixEndOption),
            MaxConcurrentActivityPollers: parseResult.GetValueForOption(MaxConcurrentActivityPollersOption),
            MaxConcurrentWorkflowPollers: parseResult.GetValueForOption(MaxConcurrentWorkflowPollersOption),
            ActivityPollerAutoscaleMax: parseResult.GetValueForOption(ActivityPollerAutoscaleMaxOption),
            WorkflowPollerAutoscaleMax: parseResult.GetValueForOption(WorkflowPollerAutoscaleMaxOption),
            MaxConcurrentActivities: parseResult.GetValueForOption(MaxConcurrentActivitiesOption),
            MaxConcurrentWorkflowTasks: parseResult.GetValueForOption(MaxConcurrentWorkflowTasksOption),
            ActivitiesPerSecond: parseResult.GetValueForOption(ActivitiesPerSecondOption),
            LogLevel: parseResult.GetValueForOption(LogLevelOption)!,
            LogEncoding: parseResult.GetValueForOption(LogEncodingOption)!,
            Tls: parseResult.GetValueForOption(TlsOption),
            TlsCertPath: parseResult.GetValueForOption(TlsCertPathOption)!,
            TlsKeyPath: parseResult.GetValueForOption(TlsKeyPathOption)!,
            PromListenAddress: parseResult.GetValueForOption(PromListenAddressOption),
            AuthHeader: parseResult.GetValueForOption(AuthHeaderOption)!,
            PromHandlerPath: parseResult.GetValueForOption(PromHandlerPathOption),
            BuildId: parseResult.GetValueForOption(BuildIdOption),
            ErrOnUnimplemented: parseResult.GetValueForOption(ErrOnUnimplementedOption));

    internal static RootCommand CreateWorkerCommand()
    {
        var command = new RootCommand(".NET project harness worker");
        command.AddOption(ServerAddressOption);
        command.AddOption(TaskQueueOption);
        command.AddOption(NamespaceOption);
        command.AddOption(TaskQueueSuffixStartOption);
        command.AddOption(TaskQueueSuffixEndOption);
        command.AddOption(MaxConcurrentActivityPollersOption);
        command.AddOption(MaxConcurrentWorkflowPollersOption);
        command.AddOption(ActivityPollerAutoscaleMaxOption);
        command.AddOption(WorkflowPollerAutoscaleMaxOption);
        command.AddOption(MaxConcurrentActivitiesOption);
        command.AddOption(MaxConcurrentWorkflowTasksOption);
        command.AddOption(ActivitiesPerSecondOption);
        command.AddOption(LogLevelOption);
        command.AddOption(LogEncodingOption);
        command.AddOption(TlsOption);
        command.AddOption(TlsCertPathOption);
        command.AddOption(TlsKeyPathOption);
        command.AddOption(PromListenAddressOption);
        command.AddOption(AuthHeaderOption);
        command.AddOption(PromHandlerPathOption);
        command.AddOption(BuildIdOption);
        command.AddOption(ErrOnUnimplementedOption);
        return command;
    }

    private sealed record WorkerSettings(
        int? MaxConcurrentActivities,
        int? MaxConcurrentWorkflowTasks,
        double? ActivitiesPerSecond,
        int? MaxConcurrentActivityPollers,
        int? MaxConcurrentWorkflowPollers,
        int? ActivityPollerAutoscaleMax,
        int? WorkflowPollerAutoscaleMax);

    internal sealed record WorkerCliOptions(
        string ServerAddress,
        string TaskQueue,
        string Namespace,
        int TaskQueueSuffixIndexStart,
        int TaskQueueSuffixIndexEnd,
        int? MaxConcurrentActivityPollers,
        int? MaxConcurrentWorkflowPollers,
        int? ActivityPollerAutoscaleMax,
        int? WorkflowPollerAutoscaleMax,
        int? MaxConcurrentActivities,
        int? MaxConcurrentWorkflowTasks,
        double? ActivitiesPerSecond,
        string LogLevel,
        string LogEncoding,
        bool Tls,
        string TlsCertPath,
        string TlsKeyPath,
        string? PromListenAddress,
        string AuthHeader,
        string? PromHandlerPath,
        string? BuildId,
        bool ErrOnUnimplemented);
}

File: workers/ruby/projects/harness/worker.rb
# frozen_string_literal: true

require 'optparse'
require 'temporalio/worker'
require_relative 'client'
require_relative 'helpers'

module Harness
  WorkerContext = Data.define(
    :logger,
    :task_queue,
    :err_on_unimplemented,
    :worker_kwargs
  )

  module WorkerCLI
    module_function

    def run_cli(worker_factory, client_factory, argv)
      run(worker_factory, client_factory, argv)
    end

    def run(worker_factory, client_factory, argv)
      options = default_options
      build_parser(options).parse!(Array(argv).dup)

      if options[:task_queue_suffix_index_start] > options[:task_queue_suffix_index_end]
        raise ArgumentError,
              'Task queue suffix start after end'
      end

      logger = Helpers.configure_logger(options[:log_level], options[:log_encoding])
      config = ClientHelpers.build_client_config(
        server_address: options[:server_address],
        namespace: options[:namespace],
        auth_header: options[:auth_header],
        tls: options[:tls],
        tls_cert_path: options[:tls_cert_path],
        tls_key_path: options[:tls_key_path],
        prom_listen_address: options[:prom_listen_address]
      )
      client = client_factory.call(config)

      task_queues = build_task_queues(
        logger,
        options[:task_queue],
        options[:task_queue_suffix_index_start],
        options[:task_queue_suffix_index_end]
      )
      worker_kwargs = build_worker_kwargs(options)
      workers = task_queues.map do |task_queue|
        worker_factory.call(
          client,
          WorkerContext.new(
            logger: logger,
            task_queue: task_queue,
            err_on_unimplemented: options[:err_on_unimplemented],
            worker_kwargs: worker_kwargs
          )
        )
      end
      run_workers(workers)
    end

    def run_workers(workers)
      # The Ruby SDK already owns/supports coordinated multi-worker shutdown.
      Temporalio::Worker.run_all(*workers, shutdown_signals: ['SIGINT'])
    end

    def build_parser(options)
      OptionParser.new do |opts|
        opts.banner = 'Usage: runner.rb [options]'

        opts.on('-q', '--task-queue QUEUE', 'Task queue to use') { |value| options[:task_queue] = value }
        opts.on('--task-queue-suffix-index-start N', Integer) do |value|
          options[:task_queue_suffix_index_start] = value
        end
        opts.on('--task-queue-suffix-index-end N', Integer) { |value| options[:task_queue_suffix_index_end] = value }
        opts.on('--max-concurrent-activity-pollers N', Integer) do |value|
          options[:max_concurrent_activity_pollers] = value
        end
        opts.on('--max-concurrent-workflow-pollers N', Integer) do |value|
          options[:max_concurrent_workflow_pollers] = value
        end
        opts.on('--activity-poller-autoscale-max N', Integer) do |value|
          options[:activity_poller_autoscale_max] = value
        end
        opts.on('--workflow-poller-autoscale-max N', Integer) do |value|
          options[:workflow_poller_autoscale_max] = value
        end
        opts.on('--max-concurrent-activities N', Integer) { |value| options[:max_concurrent_activities] = value }
        opts.on('--max-concurrent-workflow-tasks N', Integer) do |value|
          options[:max_concurrent_workflow_tasks] = value
        end
        opts.on('--activities-per-second N', Float) { |value| options[:activities_per_second] = value }
        opts.on('--err-on-unimplemented BOOL') { |value| options[:err_on_unimplemented] = parse_bool(value) }
        opts.on('--log-level LEVEL') { |value| options[:log_level] = value }
        opts.on('--log-encoding ENCODING') { |value| options[:log_encoding] = value }
        opts.on('-n', '--namespace NAMESPACE') { |value| options[:namespace] = value }
        opts.on('-a', '--server-address ADDRESS') { |value| options[:server_address] = value }
        opts.on('--tls BOOL') { |value| options[:tls] = parse_bool(value) }
        opts.on('--tls-cert-path PATH') { |value| options[:tls_cert_path] = value }
        opts.on('--tls-key-path PATH') { |value| options[:tls_key_path] = value }
        opts.on('--prom-listen-address ADDRESS') { |value| options[:prom_listen_address] = value }
        opts.on('--prom-handler-path PATH') { |_value| nil }
        opts.on('--auth-header HEADER') { |value| options[:auth_header] = value }
        opts.on('--build-id ID') { |_value| nil }
      end
    end

    def build_task_queues(logger, task_queue, suffix_start, suffix_end)
      if suffix_end.zero?
        logger.info("Ruby worker will run on task queue #{task_queue}")
        return [task_queue]
      end

      task_queues = (suffix_start..suffix_end).map { |index| "#{task_queue}-#{index}" }
      logger.info("Ruby worker will run on #{task_queues.length} task queue(s)")
      task_queues
    end

    def build_worker_kwargs(options)
      worker_kwargs = {}
      if options[:activity_poller_autoscale_max]
        worker_kwargs[:activity_task_poller_behavior] = Temporalio::Worker::PollerBehavior::Autoscaling.new(
          maximum: options[:activity_poller_autoscale_max]
        )
      elsif options[:max_concurrent_activity_pollers]
        worker_kwargs[:max_concurrent_activity_task_polls] = options[:max_concurrent_activity_pollers]
      end

      if options[:workflow_poller_autoscale_max]
        worker_kwargs[:workflow_task_poller_behavior] = Temporalio::Worker::PollerBehavior::Autoscaling.new(
          maximum: options[:workflow_poller_autoscale_max]
        )
      elsif options[:max_concurrent_workflow_pollers]
        worker_kwargs[:max_concurrent_workflow_task_polls] = options[:max_concurrent_workflow_pollers]
      end

      if options[:max_concurrent_activities]
        worker_kwargs[:max_concurrent_activities] =
          options[:max_concurrent_activities]
      end
      if options[:max_concurrent_workflow_tasks]
        worker_kwargs[:max_concurrent_workflow_tasks] =
          options[:max_concurrent_workflow_tasks]
      end
      worker_kwargs[:max_activities_per_second] = options[:activities_per_second] if options[:activities_per_second]
      worker_kwargs
    end

    def parse_bool(value)
      return value if [true, false].include?(value)

      %w[true 1 yes].include?(value.to_s.downcase)
    end

    def default_options
      {
        task_queue: 'omes',
        task_queue_suffix_index_start: 0,
        task_queue_suffix_index_end: 0,
        max_concurrent_activity_pollers: nil,
        max_concurrent_workflow_pollers: nil,
        activity_poller_autoscale_max: nil,
        workflow_poller_autoscale_max: nil,
        max_concurrent_activities: nil,
        max_concurrent_workflow_tasks: nil,
        activities_per_second: nil,
        err_on_unimplemented: false,
        log_level: 'info',
        log_encoding: 'console',
        namespace: 'default',
        server_address: 'localhost:7233',
        tls: false,
        tls_cert_path: '',
        tls_key_path: '',
        prom_listen_address: nil,
        auth_header: ''
      }
    end
  end
end
